File: forge/lib/run-state.mjs
import { createHash, randomUUID } from "node:crypto";
import {
	closeSync,
	existsSync,
	fsyncSync,
	mkdirSync,
	openSync,
	readFileSync,
	renameSync,
	rmSync,
	writeFileSync,
} from "node:fs";
import { hostname } from "node:os";
import { basename, dirname, join } from "node:path";

export const RUN_STATE_SCHEMA_VERSION = 1;
export const DEFAULT_MAX_ATTEMPTS = 3;

function canonicalValue(value) {
	if (Array.isArray(value)) return value.map(canonicalValue);
	if (value && typeof value === "object") {
		return Object.fromEntries(Object.entries(value).sort(([left], [right]) => left.localeCompare(right)).map(([key, item]) => [key, canonicalValue(item)]));
	}
	return value;
}

export function canonicalJson(value) {
	return JSON.stringify(canonicalValue(value));
}

export function configurationFingerprint(value) {
	return createHash("sha256").update(canonicalJson(value)).digest("hex");
}

function fsyncDirectory(directory) {
	let descriptor;
	try {
		descriptor = openSync(directory, "r");
		fsyncSync(descriptor);
	} catch {
		// Some filesystems do not permit directory fsync. The file itself was synced.
	} finally {
		if (descriptor !== undefined) closeSync(descriptor);
	}
}

export function atomicWriteFile(filePath, value) {
	mkdirSync(dirname(filePath), { recursive: true });
	const temporaryPath = join(dirname(filePath), `.${basename(filePath)}.${process.pid}.${randomUUID()}.tmp`);
	const descriptor = openSync(temporaryPath, "wx");
	try {
		writeFileSync(descriptor, value);
		fsyncSync(descriptor);
	} finally {
		closeSync(descriptor);
	}
	renameSync(temporaryPath, filePath);
	fsyncDirectory(dirname(filePath));
}

export function atomicWriteJson(filePath, value) {
	atomicWriteFile(filePath, `${JSON.stringify(value, null, 2)}\n`);
}

export function appendJsonlFsync(filePath, value) {
	mkdirSync(dirname(filePath), { recursive: true });
	const descriptor = openSync(filePath, "a");
	try {
		writeFileSync(descriptor, `${JSON.stringify(value)}\n`);
		fsyncSync(descriptor);
	} finally {
		closeSync(descriptor);
	}
}

export function readJsonlRecoverTail(filePath, { repair = false } = {}) {
	if (!existsSync(filePath)) return { rows: [], warnings: [] };
	const text = readFileSync(filePath, "utf8");
	const lines = text.split("\n");
	const rows = [];
	const warnings = [];
	const validLines = [];
	for (let index = 0; index < lines.length; index += 1) {
		const line = lines[index];
		if (!line.trim()) continue;
		try {
			rows.push(JSON.parse(line));
			validLines.push(line);
		} catch (error) {
			const isTail = index === lines.length - 1 && !text.endsWith("\n");
			if (!isTail) throw new Error(`invalid JSONL at ${filePath}:${index + 1}: ${error instanceof Error ? error.message : String(error)}`);
			warnings.push(`Ignored an incomplete final JSONL record at ${filePath}:${index + 1}.`);
			if (repair) atomicWriteFile(filePath, validLines.length > 0 ? `${validLines.join("\n")}\n` : "");
		}
	}
	return { rows, warnings };
}

function now() {
	return new Date().toISOString();
}

export function createRunState({ workflow, command, input, options, items = [], phase = "initialized", nextAction = null, children = {} }) {
	const createdAt = now();
	return {
		schemaVersion: RUN_STATE_SCHEMA_VERSION,
		workflow,
		createdAt,
		updatedAt: createdAt,
		command,
		input,
		options,
		optionsFingerprint: configurationFingerprint({ workflow, command, input, options }),
		status: "running",
		phase,
		nextAction,
		items,
		children,
		warnings: [],
	};
}

export function initializeRunState(runDirectory, state) {
	if (existsSync(join(runDirectory, "run_state.json"))) throw new Error(`run state already exists: ${runDirectory}`);
	atomicWriteJson(join(runDirectory, "run_state.json"), state);
	appendRunEvent(runDirectory, { type: "run_initialized", workflow: state.workflow, phase: state.phase });
	return state;
}

export function loadRunState(runDirectory, workflow = null) {
	const statePath = join(runDirectory, "run_state.json");
	if (!existsSync(statePath)) throw new Error(`legacy or unrelated output directory has no run_state.json: ${runDirectory}`);
	const state = JSON.parse(readFileSync(statePath, "utf8"));
	if (state.schemaVersion !== RUN_STATE_SCHEMA_VERSION) throw new Error(`unsupported run state schema version: ${state.schemaVersion}`);
	if (workflow && state.workflow !== workflow) throw new Error(`run belongs to ${state.workflow}, not ${workflow}`);
	return state;
}

export function assertCompatibleRun(state, configuration) {
	const actual = configurationFingerprint(configuration);
	if (actual !== state.optionsFingerprint) throw new Error("existing run options or input do not match this invocation; use status/refresh or choose a new output directory");
}

export function updateRunState(runDirectory, mutate, event = null) {
	const state = loadRunState(runDirectory);
	const draft = structuredClone(state);
	const updated = mutate(draft) ?? draft;
	updated.updatedAt = now();
	atomicWriteJson(join(runDirectory, "run_state.json"), updated);
	if (event) appendRunEvent(runDirectory, event);
	return updated;
}

export function appendRunEvent(runDirectory, event) {
	const eventPath = join(runDirectory, "run_events.jsonl");
	const prior = readJsonlRecoverTail(eventPath, { repair: true }).rows;
	appendJsonlFsync(eventPath, { sequence: prior.length + 1, at: now(), ...event });
}

export function inputDrift(snapshot, current) {
	const original = new Map(snapshot.map((item) => [item.path, item]));
	const observed = new Map(current.map((item) => [item.path, item]));
	const added = current.filter((item) => !original.has(item.path));
	const removed = snapshot.filter((item) => !observed.has(item.path));
	const changed = current.filter((item) => original.has(item.path) && original.get(item.path).sha256 !== item.sha256);
	return { changed: changed.map((item) => ({ before: original.get(item.path), after: item })), added, removed };
}

export function isTransientFailure(error) {
	const code = String(error?.code ?? "").toLowerCase();
	const message = String(error instanceof Error ? error.message : error).toLowerCase();
	return ["econnreset", "econnrefused", "etimedout", "timeout", "interrupted", "aborted"].some((value) => code.includes(value) || message.includes(value)) || /http\s+5\d\d/.test(message);
}

export function retryableItem(item, maximumAttempts = DEFAULT_MAX_ATTEMPTS) {
	return item.status === "pending" || item.status === "in_progress" || (item.status === "failed" && item.transient === true && (item.attempts ?? 0) < maximumAttempts);
}

function processIsAlive(pid) {
	if (!Number.isInteger(pid) || pid < 1) return false;
	try {
		process.kill(pid, 0);
		return true;
	} catch (error) {
		return error?.code === "EPERM";
	}
}

export async function withRunLock(runDirectory, callback) {
	const lockPath = join(runDirectory, ".run.lock");
	const acquire = () => {
		try {
			writeFileSync(lockPath, `${JSON.stringify({ pid: process.pid, host: hostname(), createdAt: now() })}\n`, { flag: "wx" });
			return;
		} catch (error) {
			if (error?.code !== "EEXIST") throw error;
		}
		let lock = null;
		try {
			lock = JSON.parse(readFileSync(lockPath, "utf8"));
		} catch {
			// A malformed lock is stale.
		}
		if (lock?.host && lock.host !== hostname()) throw new Error(`run is locked by PID ${lock.pid ?? "unknown"} on ${lock.host}`);
		if (processIsAlive(lock?.pid)) throw new Error(`run is locked by active PID ${lock.pid}`);
		rmSync(lockPath, { force: true });
		writeFileSync(lockPath, `${JSON.stringify({ pid: process.pid, host: hostname(), createdAt: now() })}\n`, { flag: "wx" });
	};
	mkdirSync(runDirectory, { recursive: true });
	acquire();
	try {
		return await callback();
	} finally {
		rmSync(lockPath, { force: true });
	}
}

File: forge/lib/run_state.py
#!/usr/bin/env python3

import contextlib
import hashlib
import json
import os
import socket
import tempfile
from datetime import datetime, timezone
from pathlib import Path


RUN_STATE_SCHEMA_VERSION = 1
DEFAULT_MAX_ATTEMPTS = 3


def utc_now():
    return datetime.now(timezone.utc).isoformat()


def canonical_json(value):
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def configuration_fingerprint(value):
    return hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()


def atomic_write_text(path, value):
    destination = Path(path)
    destination.parent.mkdir(parents=True, exist_ok=True)
    descriptor, temporary = tempfile.mkstemp(prefix=f".{destination.name}.", suffix=".tmp", dir=destination.parent)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8", newline="") as handle:
            handle.write(value)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary, destination)
        try:
            directory_descriptor = os.open(destination.parent, os.O_RDONLY)
            try:
                os.fsync(directory_descriptor)
            finally:
                os.close(directory_descriptor)
        except OSError:
            pass
    except BaseException:
        with contextlib.suppress(FileNotFoundError):
            os.unlink(temporary)
        raise


def atomic_write_json(path, value):
    atomic_write_text(path, json.dumps(value, indent=2, ensure_ascii=False) + "\n")


def append_jsonl_fsync(path, value):
    destination = Path(path)
    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("a", encoding="utf-8", newline="") as handle:
        handle.write(json.dumps(value, ensure_ascii=False) + "\n")
        handle.flush()
        os.fsync(handle.fileno())


def read_jsonl_recover_tail(path, repair=False):
    source = Path(path)
    if not source.is_file():
        return [], []
    text = source.read_text(encoding="utf-8")
    lines = text.splitlines(keepends=True)
    rows = []
    valid = []
    warnings = []
    for index, raw in enumerate(lines, start=1):
        line = raw.strip()
        if not line:
            continue
        try:
            rows.append(json.loads(line))
            valid.append(line)
        except json.JSONDecodeError as error:
            is_tail = index == len(lines) and not raw.endswith("\n")
            if not is_tail:
                raise ValueError(f"invalid JSONL at {source}:{index}: {error}") from error
            warnings.append(f"Ignored an incomplete final JSONL record at {source}:{index}.")
            if repair:
                atomic_write_text(source, "" if not valid else "\n".join(valid) + "\n")
    return rows, warnings


def create_run_state(workflow, command, input_config, options, items=None, phase="initialized", next_action=None, children=None):
    created_at = utc_now()
    configuration = {"workflow": workflow, "command": command, "input": input_config, "options": options}
    return {
        "schemaVersion": RUN_STATE_SCHEMA_VERSION,
        "workflow": workflow,
        "createdAt": created_at,
        "updatedAt": created_at,
        "command": command,
        "input": input_config,
        "options": options,
        "optionsFingerprint": configuration_fingerprint(configuration),
        "status": "running",
        "phase": phase,
        "nextAction": next_action,
        "items": items or [],
        "children": children or {},
        "warnings": [],
    }


def initialize_run_state(run_directory, state):
    root = Path(run_directory)
    state_path = root / "run_state.json"
    if state_path.exists():
        raise ValueError(f"run state already exists: {root}")
    atomic_write_json(state_path, state)
    append_run_event(root, {"type": "run_initialized", "workflow": state["workflow"], "phase": state["phase"]})
    return state


def load_run_state(run_directory, workflow=None):
    root = Path(run_directory)
    state_path = root / "run_state.json"
    if not state_path.is_file():
        raise ValueError(f"legacy or unrelated output directory has no run_state.json: {root}")
    state = json.loads(state_path.read_text(encoding="utf-8"))
    if state.get("schemaVersion") != RUN_STATE_SCHEMA_VERSION:
        raise ValueError(f"unsupported run state schema version: {state.get('schemaVersion')}")
    if workflow and state.get("workflow") != workflow:
        raise ValueError(f"run belongs to {state.get('workflow')}, not {workflow}")
    return state


def assert_compatible_run(state, configuration):
    if configuration_fingerprint(configuration) != state.get("optionsFingerprint"):
        raise ValueError("existing run options or input do not match this invocation; use status/refresh or choose a new output directory")


def update_run_state(run_directory, mutate, event=None):
    root = Path(run_directory)
    state = load_run_state(root)
    draft = json.loads(json.dumps(state))
    updated = mutate(draft) or draft
    updated["updatedAt"] = utc_now()
    atomic_write_json(root / "run_state.json", updated)
    if event:
        append_run_event(root, event)
    return updated


def append_run_event(run_directory, event):
    root = Path(run_directory)
    path = root / "run_events.jsonl"
    prior, _ = read_jsonl_recover_tail(path, repair=True)
    append_jsonl_fsync(path, {"sequence": len(prior) + 1, "at": utc_now(), **event})


def input_drift(snapshot, current):
    original = {item["path"]: item for item in snapshot}
    observed = {item["path"]: item for item in current}
    return {
        "added": [item for item in current if item["path"] not in original],
        "removed": [item for item in snapshot if item["path"] not in observed],
        "changed": [
            {"before": original[item["path"]], "after": item}
            for item in current
            if item["path"] in original and original[item["path"]].get("sha256") != item.get("sha256")
        ],
    }


def is_transient_failure(error):
    code = str(getattr(error, "code", "")).lower()
    message = str(error).lower()
    values = ("econnreset", "econnrefused", "etimedout", "timeout", "interrupted", "aborted")
    return any(value in code or value in message for value in values) or any(f"http {status}" in message for status in range(500, 600))


def retryable_item(item, maximum_attempts=DEFAULT_MAX_ATTEMPTS):
    return item.get("status") in {"pending", "in_progress"} or (
        item.get("status") == "failed" and item.get("transient") is True and item.get("attempts", 0) < maximum_attempts
    )


def _pid_alive(pid):
    if not isinstance(pid, int) or pid < 1:
        return False
    try:
        os.kill(pid, 0)
        return True
    except PermissionError:
        return True
    except ProcessLookupError:
        return False


@contextlib.contextmanager
def run_lock(run_directory):
    root = Path(run_directory)
    root.mkdir(parents=True, exist_ok=True)
    lock_path = root / ".run.lock"
    payload = {"pid": os.getpid(), "host": socket.gethostname(), "createdAt": utc_now()}
    try:
        descriptor = os.open(lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    except FileExistsError:
        try:
            existing = json.loads(lock_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            existing = {}
        if existing.get("host") and existing["host"] != socket.gethostname():
            raise RuntimeError(f"run is locked by PID {existing.get('pid', 'unknown')} on {existing['host']}") from None
        if _pid_alive(existing.get("pid")):
            raise RuntimeError(f"run is locked by active PID {existing['pid']}") from None
        lock_path.unlink(missing_ok=True)
        descriptor = os.open(lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(descriptor, "w", encoding="utf-8") as handle:
        handle.write(json.dumps(payload) + "\n")
        handle.flush()
        os.fsync(handle.fileno())
    try:
        yield
    finally:
        lock_path.unlink(missing_ok=True)
